packages/react-server-dom-webpack/src/__tests__/ReactFlightDOMEdge-test.js

```diff
@@ -543,6 +543,53 @@ describe('ReactFlightDOMEdge', () => {
     expect(await iterator.next()).toEqual({value: undefined, done: true});
   });
 
+  it('dedupes objects inside async iterables or readablestreams', async () => {
+    const obj = {
+      this: {is: 'a large objected'},
+      with: {many: 'properties in it'},
+    };
+    const iterable = {
+      async *[Symbol.asyncIterator]() {
+        for (let i = 0; i < 30; i++) {
+          yield obj;
+        }
+      },
+    };
+
+    const stream = ReactServerDOMServer.renderToReadableStream({
+      iterable,
+    });
+    const [stream1, stream2] = passThrough(stream).tee();
+
+    const serializedContent = await readResult(stream1);
+    // TODO: Ideally streams should dedupe objects but because we never outline the objects
+    // they end up not having a row to reference them nor any of its nested objects.
+    // expect(serializedContent.length).toBeLessThan(400);
```

Comment:
Add a test to show case this.

Basically, we because every "row" inside a stream has the same ID we can't refer to those objects. Only the stream object itself (`ReadableStream` or `AsyncIterable`) has a row ID.

This means we can't dedupe the object at the root of a stream row. Since we don't have a reference to that object we also can't dedupe the nested object via the parent neither. Leading to duplicating all these objects.

It seems unfortunate to outline every chunk/entry in a stream just in case it needs deduping.

Reply:
Potentially there could be a way to reach into the nth item inside the buffer of the stream to get the Chunk responsible for that and then reach into it.